src/kernel/arch/aarch64/boot.zig

```diff
@@ -8,32 +9,39 @@ extern var KERNEL_STACK_END: *usize;
 
 export fn _start() linksection(".text.boot") callconv(.Naked) noreturn {
     // Halt all cores other than the primary core, until the kernel has multicore support
-    const core_id = asm ("mrs %[res], mpidr_el1"
-        : [res] "=r" (-> usize)
-    ) & 3;
-    if (core_id != 0) {
-        while (true) {
-            asm volatile ("wfe");
-        }
-    }
+    asm volatile (
+        \\ mrs x0, mpidr_el1
+        \\ mov x1, #3
+        \\ ands x0, x0, x1
+        \\ beq cpu0
+        \\hang:
+        \\ wfe
+        \\ b hang
+        \\cpu0:
+    );
```

Comment:
Raw assembly again :)

Reply:
This is necessary because the disassembly of the zig version showed the use of the sp which is not yet initialized. Each core will need its own sp so sp cannot be initialized yet.

I will add a comment.

Reply:
Oh well spotted. It's stupid use of the stack too since the stored value is loaded immediately afterwards. Perhaps you could read `mpidr_el1` in assembly then do the comparison with 0 and loop in Zig?

Reply:
Possibly, but that loop might be using the stack.

Can we keep it assembly in the pr and have an issue to improve it?

Reply:
Have a look and see if looping does end up using the stack in the debug build mode. If it does then I reckon you can keep it as it is.

Reply:
I think an idea might be to keep the `_start` function really small and only setup a temporary stack (with asm volatile) and call another function. That way the `_start` function will not use the stack and the next function can do the checking of the core number etc and afterwards setting up the correct stack. The small temporary stack only has to be like 256 bytes or something small.

Or just setup a small temporary stack as the first step in the `_start` function.

Reply:
> I think an idea might be to keep the `_start` function really small and only setup a temporary stack (with asm volatile) and call another function. That way the `_start` function will not use the stack and the next function can do the checking of the core number etc and afterwards setting up the correct stack. The small temporary stack only has to be like 256 bytes or something small.
> 
> Or just setup a small temporary stack as the first step in the `_start` function.

@FireFox317 That sounds ideal to me. The only consideration would be if the cores using the same stack would cause a conflict.

Reply:
@SamTebbs33 yeah obviously you want different stacks per core, otherwise the cores would overwrite the data of the other core. I would do something like the following:
```asm
mrs x0, mpidr_el1 // get system register
and x0, x0, #3 // mask out bits not belonging to core id
mul x1, x0, #255 // 255 bytes per stack
mov x2, <stack_address> // store the lowest stack address in x2
sub x2, x2, x1 
mov sp, x2
```
This can probably written in smaller assembly.

Reply:
@SamTebbs33 I will change to this approach.

Reply:
@SamTebbs33 @FireFox317

boot.zig: 

https://github.com/markfirmware/pluto/blob/1b04bd80a27c560a7777542e38d07f2844ff8ecd/src/kernel/arch/aarch64/boot.zig#L1-L50

cpu registers:

https://github.com/markfirmware/pluto/blob/1b04bd80a27c560a7777542e38d07f2844ff8ecd/src/kernel/arch/aarch64/arch.zig#L118-L178

Reply:
Cheers